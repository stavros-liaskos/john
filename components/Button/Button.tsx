import React, { useCallback } from 'react';
import { ButtonProps } from './Button.types';
import Spin from '../Icons/spin';

const Button: React.FunctionComponent<ButtonProps> = ({
  className,
  i18n,
  handleClick = () => {},
  disabled = false,
  loading = false,
  type = 'button',
  handleClickArg,
}) => {
  const handleClickCallback = useCallback(() => {
    handleClickArg ? handleClick(handleClickArg) : handleClick();
  }, [handleClick, handleClickArg]);

  if (!i18n) {
    return null;
  }

  return (
    <button className={`btn flex ${className}`} onClick={handleClickCallback} disabled={disabled} type={type}>
      {loading && <Spin />}
      {i18n}
    </button>
  );
};
Button.whyDidYouRender = true;
export default Button;
