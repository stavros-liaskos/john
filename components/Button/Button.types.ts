export interface ButtonProps {
  i18n: string;
  className?: string;
  handleClick?: (...args: number[]) => void;
  handleClickArg?: number;
  disabled?: boolean;
  loading?: boolean;
  type?: 'submit';
}
