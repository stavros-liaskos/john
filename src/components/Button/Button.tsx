import React, { useCallback } from 'react';
import { ButtonProps } from './Button.types';
import Spin from '../Icons/spin';

const Button: React.FunctionComponent<ButtonProps> = ({
  className,
  i18n,
  handleClick = () => {},
  disabled = false,
  loading = false,
  type = 'button',
  handleClickArg,
  children,
}) => {
  const handleClickCallback = useCallback(() => {
    if (!disabled) {
      handleClickArg ? handleClick(handleClickArg) : handleClick();
    }
  }, [handleClick, handleClickArg, disabled]);

  if (!i18n && !children) {
    return null;
  }

  return (
    <button
      className={`btn flex ${className ? className : ''}`}
      onClick={handleClickCallback}
      disabled={disabled}
      type={type}
    >
      {loading && (
        <span className="flex justify-center items-center -ml-1 mr-3 h-5 w-5">
          <Spin width={20} />
        </span>
      )}
      {i18n}
      {children}
    </button>
  );
};
Button.whyDidYouRender = true;
export default Button;
