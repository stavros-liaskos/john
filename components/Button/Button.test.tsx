import React from 'react';
import { render } from '@testing-library/react';
import Button from './Button';

describe('Button', () => {
  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<Button />);
  });

  it('renders an enabled button by default and not a loading spinner', () => {
    const { container } = render(<Button i18n="Click Me!" />);

    // TODO validate disabled attribute
    // TODO validate spinner doesn't exist
    expect(container).toMatchSnapshot();
  });

  it('renders a disabled button', () => {
    const { container } = render(<Button i18n="No Action" disabled={true} />);

    // TODO validate disabled attribute
    expect(container).toMatchSnapshot();
  });

  it('renders a spinner when loading', () => {
    const { container } = render(<Button i18n="No Action" loading={true} />);

    // TODO validate spinner exists
    expect(container).toMatchSnapshot();
  });

  it('calls to action when button clicked', () => {
    const clickHandler = jest.fn();
    const { container } = render(<Button i18n="Click Me!" loading={true} onClick={clickHandler} />);

    // TODO clickHandler have been called
    expect(container).toMatchSnapshot();
  });
});
