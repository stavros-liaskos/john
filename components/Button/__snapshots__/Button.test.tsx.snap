// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button calls to action when button clicked 1`] = `
<div>
  <button
    class="btn flex undefined"
  >
    <svg
      class="animate-spin -ml-1 mr-3 h-5 w-5 text:black dark:text-white"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
    Click Me!
  </button>
</div>
`;

exports[`Button renders a disabled button 1`] = `
<div>
  <button
    class="btn flex undefined"
    disabled=""
  >
    No Action
  </button>
</div>
`;

exports[`Button renders a spinner when loading 1`] = `
<div>
  <button
    class="btn flex undefined"
  >
    <svg
      class="animate-spin -ml-1 mr-3 h-5 w-5 text:black dark:text-white"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
    No Action
  </button>
</div>
`;

exports[`Button renders an enabled button by default and not a loading spinner 1`] = `
<div>
  <button
    class="btn flex undefined"
  >
    Click Me!
  </button>
</div>
`;
